Bottom: 3038a0f29cb30b2ae45da91054dec79d6bc2e21a
Top:    04ad3293daab7a54da5a541e0b6bbf1225438804
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-12-23 21:31:35 +0000

hid/common/hidgl.c: Simplify angle calculation for drawing line caps

Make use of atan2 to avoid special casing horizontal lines, and move
make the input variables doubles to be consistent with that function.
Our use of "tanl" was inappropriate for the float input, dobule output
variables we were using before (spotted by Dan McMahill).

I have modified the semantic meaning of angle, to keep the more
conventional angle = atan2 (dy, dx); geometry. This means the angle
now refers to the angle of the line, not the angle of the line-cap
start. (The angle is adjusted before passing into the cap drawing
routines).


---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 7ec148f..15273a6 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -215,7 +215,7 @@ void
 hidgl_draw_line (int cap, Coord width, Coord x1, Coord y1, Coord x2, Coord y2, double scale)
 {
   double angle;
-  float deltax, deltay, length;
+  double deltax, deltay, length;
   float wdx, wdy;
   int circular_caps = 0;
   int hairline = 0;
@@ -233,7 +233,6 @@ hidgl_draw_line (int cap, Coord width, Coord x1, Coord y1, Coord x2, Coord y2, d
 
   if (length == 0) {
     /* Assume the orientation of the line is horizontal */
-    angle = 0;
     wdx = -width / 2.;
     wdy = 0;
     length = 1.;
@@ -242,16 +241,10 @@ hidgl_draw_line (int cap, Coord width, Coord x1, Coord y1, Coord x2, Coord y2, d
   } else {
     wdy = deltax * width / 2. / length;
     wdx = -deltay * width / 2. / length;
-
-    if (deltay == 0.)
-      angle = (deltax < 0) ? 270. : 90.;
-    else
-      angle = 180. / M_PI * atanl (deltax / deltay);
-
-    if (deltay < 0)
-      angle += 180.;
   }
 
+  angle = -180. / M_PI * atan2 (deltay, deltax);
+
   switch (cap) {
     case Trace_Cap:
     case Round_Cap:
@@ -278,8 +271,8 @@ hidgl_draw_line (int cap, Coord width, Coord x1, Coord y1, Coord x2, Coord y2, d
   /* Don't bother capping hairlines */
   if (circular_caps && !hairline)
     {
-      draw_cap (width, x1, y1, angle, scale);
-      draw_cap (width, x2, y2, angle + 180., scale);
+      draw_cap (width, x1, y1, angle + 90., scale);
+      draw_cap (width, x2, y2, angle - 90., scale);
     }
 }
